File: modulo4/flutter_shop_app/lib/presentation/screens/orders/orders_screen.dart
// lib/presentation/screens/orders/orders_screen.dart

import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import '../../../theme/app_colors.dart';
import '../../../core/utils/formatters.dart';
import '../../../domain/model/order.dart';
import '../../providers/orders_provider.dart';
import '../../widgets/status_badge.dart';

const _statusFilters = [
  ('',          'Todos'),
  ('pending',   'Pendientes'),
  ('confirmed', 'Confirmados'),
  ('shipped',   'Enviados'),
  ('delivered', 'Entregados'),
  ('cancelled', 'Cancelados'),
];

class OrdersScreen extends ConsumerStatefulWidget {
  const OrdersScreen({super.key});

  @override
  ConsumerState<OrdersScreen> createState() => _OrdersScreenState();
}

class _OrdersScreenState extends ConsumerState<OrdersScreen> {
  final _scrollCtrl = ScrollController();

  @override
  void initState() {
    super.initState();
    _scrollCtrl.addListener(() {
      if (_scrollCtrl.position.pixels >=
          _scrollCtrl.position.maxScrollExtent - 150) {
        ref.read(ordersProvider.notifier).loadMore();
      }
    });
  }

  @override
  void dispose() {
    _scrollCtrl.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final state = ref.watch(ordersProvider);
    final tt    = Theme.of(context).textTheme;

    return Scaffold(
      body: SafeArea(
        child: Column(
          children: [
            // ── Header ──────────────────────────────────────
            Container(
              color:   AppColors.surface,
              padding: const EdgeInsets.fromLTRB(20, 20, 20, 0),
              child:   Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Text('Mis pedidos', style: tt.headlineMedium),
                          Text(
                            '${state.total} pedido${state.total != 1 ? "s" : ""}',
                            style: tt.bodySmall,
                          ),
                        ],
                      ),
                      IconButton(
                        onPressed: ref.read(ordersProvider.notifier).refresh,
                        icon: const Icon(Icons.refresh_rounded, color: AppColors.textSecondary),
                      ),
                    ],
                  ),
                  const SizedBox(height: 12),

                  // Filtros por estado
                  SizedBox(
                    height: 34,
                    child:  ListView(
                      scrollDirection:  Axis.horizontal,
                      children: _statusFilters.map((filter) {
                        final isSelected = state.statusFilter == filter.$1;
                        return Padding(
                          padding: const EdgeInsets.only(right: 8),
                          child:   ChoiceChip(
                            label:     Text(filter.$2),
                            selected:  isSelected,
                            onSelected:(_) =>
                                ref.read(ordersProvider.notifier).setStatusFilter(filter.$1),
                          ),
                        );
                      }).toList(),
                    ),
                  ),
                  const SizedBox(height: 12),
                ],
              ),
            ),

            // ── Contenido ─────────────────────────────────────
            Expanded(
              child: Builder(builder: (_) {
                if (state.isLoading && state.orders.isEmpty) {
                  return const Center(
                    child: CircularProgressIndicator(color: AppColors.accent),
                  );
                }
                if (state.error != null && state.orders.isEmpty) {
                  return Center(
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        const Text('❌', style: TextStyle(fontSize: 40)),
                        const SizedBox(height: 12),
                        Text(state.error!,
                            style: const TextStyle(color: AppColors.error)),
                        const SizedBox(height: 16),
                        ElevatedButton(
                          onPressed: ref.read(ordersProvider.notifier).refresh,
                          child: const Text('Reintentar'),
                        ),
                      ],
                    ),
                  );
                }
                if (state.orders.isEmpty) {
                  return const Center(
                    child: Column(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Text('📦', style: TextStyle(fontSize: 52)),
                        SizedBox(height: 16),
                        Text('Sin pedidos',
                            style: TextStyle(
                              color: AppColors.textPrimary,
                              fontSize: 18,
                              fontWeight: FontWeight.bold,
                            )),
                        Text('Tus pedidos aparecerán aquí',
                            style: TextStyle(color: AppColors.textSecondary)),
                      ],
                    ),
                  );
                }

                return ListView.separated(
                  controller:      _scrollCtrl,
                  padding:         const EdgeInsets.all(16),
                  itemCount:       state.orders.length + (state.isLoadingMore ? 1 : 0),
                  separatorBuilder:(_, __) => const SizedBox(height: 12),
                  itemBuilder: (_, i) {
                    if (i >= state.orders.length) {
                      return const Center(
                        child: Padding(
                          padding: EdgeInsets.all(16),
                          child:   CircularProgressIndicator(
                            color: AppColors.accent, strokeWidth: 2,
                          ),
                        ),
                      );
                    }
                    final order = state.orders[i];
                    return _OrderCard(
                      order:   order,
                      onTap:   () => context.push('/orders/${order.id}'),
                    );
                  },
                );
              }),
            ),
          ],
        ),
      ),
    );
  }
}

// ── OrderCard ─────────────────────────────────────────────────

class _OrderCard extends StatelessWidget {
  final Order        order;
  final VoidCallback onTap;
  const _OrderCard({required this.order, required this.onTap});

  @override
  Widget build(BuildContext context) {
    final tt      = Theme.of(context).textTheme;
    final dateStr = formatDate(order.createdAt);

    return GestureDetector(
      onTap: onTap,
      child: Container(
        padding:    const EdgeInsets.all(16),
        decoration: BoxDecoration(
          color:        AppColors.surface,
          borderRadius: BorderRadius.circular(16),
          border:       Border.all(color: AppColors.border),
        ),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            // Header
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Text('Pedido #${order.id}', style: tt.titleMedium),
                    Text(dateStr, style: tt.bodySmall),
                  ],
                ),
                StatusBadge(status: order.status),
              ],
            ),
            const SizedBox(height: 12),

            // Preview de ítems
            Wrap(
              spacing: 6, runSpacing: 4,
              children: [
                ...order.items.take(3).map((item) => Container(
                  padding:    const EdgeInsets.symmetric(horizontal: 8, vertical: 3),
                  decoration: BoxDecoration(
                    color:        AppColors.surface2,
                    borderRadius: BorderRadius.circular(6),
                    border:       Border.all(color: AppColors.borderLight),
                  ),
                  child: Text(
                    '${item.quantity}× ${item.productName}',
                    style: const TextStyle(color: AppColors.textSecondary, fontSize: 11),
                    maxLines: 1, overflow: TextOverflow.ellipsis,
                  ),
                )),
                if (order.items.length > 3)
                  Text(
                    '+${order.items.length - 3} más',
                    style: const TextStyle(color: AppColors.textFaint, fontSize: 11),
                  ),
              ],
            ),
            const SizedBox(height: 12),
            const Divider(height: 1),
            const SizedBox(height: 10),

            // Footer
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceBetween,
              children: [
                Text(
                  '${order.numItems} producto${order.numItems != 1 ? "s" : ""}',
                  style: tt.bodySmall,
                ),
                Row(
                  children: [
                    Text(
                      formatPrice(order.total),
                      style: const TextStyle(
                        color: AppColors.accent, fontSize: 16, fontWeight: FontWeight.bold,
                      ),
                    ),
                    const SizedBox(width: 4),
                    const Icon(Icons.chevron_right, color: AppColors.textFaint, size: 18),
                  ],
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

File: modulo2/shopapp/app/src/main/java/com/shopapp/presentation/ui/client/orders/OrdersScreen.kt
// presentation/ui/client/orders/OrdersScreen.kt
package com.shopapp.presentation.ui.client.orders

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Refresh
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import com.shopapp.domain.model.Order
import com.shopapp.domain.model.OrderStatus
import com.shopapp.presentation.components.StatusBadge
import com.shopapp.presentation.components.LoadingScreen
import com.shopapp.presentation.components.ErrorScreen
import com.shopapp.presentation.viewmodel.OrdersViewModel
import com.shopapp.theme.*
import java.text.SimpleDateFormat
import java.util.Locale

private val STATUS_FILTERS = listOf(
    "" to "Todos",
    OrderStatus.PENDING.value   to "Pendientes",
    OrderStatus.CONFIRMED.value to "Confirmados",
    OrderStatus.SHIPPED.value   to "Enviados",
    OrderStatus.DELIVERED.value to "Entregados",
    OrderStatus.CANCELLED.value to "Cancelados",
)

@Composable
fun OrdersScreen(
    onOrderClick: (Int) -> Unit,
    viewModel:    OrdersViewModel = hiltViewModel(),
) {
    val state     by viewModel.state.collectAsState()
    val listState  = rememberLazyListState()

    // Cargar más al llegar al final
    val shouldLoadMore by remember {
        derivedStateOf {
            val lastVisible = listState.layoutInfo.visibleItemsInfo.lastOrNull()?.index ?: 0
            val total       = listState.layoutInfo.totalItemsCount
            lastVisible >= total - 2 && !state.isLoadingMore && state.hasMore
        }
    }
    LaunchedEffect(shouldLoadMore) {
        if (shouldLoadMore) viewModel.loadMore()
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Background),
    ) {
        // ── Header ────────────────────────────────────────────
        Surface(color = Surface, tonalElevation = 0.dp) {
            Column(modifier = Modifier.padding(16.dp)) {
                Row(
                    modifier              = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment     = Alignment.CenterVertically,
                ) {
                    Column {
                        Text(
                            text       = "Mis pedidos",
                            style      = MaterialTheme.typography.headlineMedium,
                            fontWeight = FontWeight.Bold,
                            color      = TextPrimary,
                        )
                        Text(
                            text  = if (state.isLoading) "..." else "${state.total} pedidos",
                            style = MaterialTheme.typography.bodySmall,
                            color = TextSecondary,
                        )
                    }
                    IconButton(onClick = viewModel::refresh) {
                        Icon(Icons.Default.Refresh, contentDescription = "Actualizar", tint = TextSecondary)
                    }
                }

                Spacer(Modifier.height(12.dp))

                // Filtros por estado
                LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                    items(STATUS_FILTERS) { (value, label) ->
                        FilterChip(
                            selected = state.statusFilter == value,
                            onClick  = { viewModel.setStatusFilter(value) },
                            label    = { Text(label, style = MaterialTheme.typography.labelSmall) },
                            colors   = FilterChipDefaults.filterChipColors(
                                selectedContainerColor = Accent,
                                selectedLabelColor     = AccentOnDark,
                                containerColor         = Surface2,
                                labelColor             = TextSecondary,
                            ),
                        )
                    }
                }
            }
        }

        // ── Contenido ─────────────────────────────────────────
        when {
            state.isLoading && state.orders.isEmpty() ->
                LoadingScreen("Cargando pedidos...")

            state.error != null && state.orders.isEmpty() ->
                ErrorScreen(state.error!!, onRetry = viewModel::refresh)

            state.orders.isEmpty() -> {
                Box(Modifier.fillMaxSize(), contentAlignment = Alignment.Center) {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text("📦", fontSize = 52.sp)
                        Spacer(Modifier.height(12.dp))
                        Text(
                            text       = if (state.statusFilter.isBlank()) "Sin pedidos"
                            else "Sin pedidos con este estado",
                            style      = MaterialTheme.typography.titleMedium,
                            fontWeight = FontWeight.Bold,
                            color      = TextPrimary,
                        )
                        Text(
                            text  = "Tus pedidos aparecerán aquí",
                            style = MaterialTheme.typography.bodySmall,
                            color = TextSecondary,
                        )
                    }
                }
            }

            else -> {
                LazyColumn(
                    state          = listState,
                    modifier       = Modifier.fillMaxSize(),
                    contentPadding = PaddingValues(16.dp),
                    verticalArrangement = Arrangement.spacedBy(12.dp),
                ) {
                    items(state.orders, key = { it.id }) { order ->
                        OrderCard(order = order, onClick = { onOrderClick(order.id) })
                    }

                    if (state.isLoadingMore) {
                        item {
                            Box(
                                modifier         = Modifier.fillMaxWidth().padding(16.dp),
                                contentAlignment = Alignment.Center,
                            ) {
                                CircularProgressIndicator(
                                    color       = Accent,
                                    modifier    = Modifier.size(28.dp),
                                    strokeWidth = 2.dp,
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

// ── OrderCard ─────────────────────────────────────────────────

@Composable
fun OrderCard(order: Order, onClick: () -> Unit) {
    val inputFmt  = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", Locale.getDefault())
    val outputFmt = SimpleDateFormat("dd MMM yyyy", Locale("es"))
    val dateStr   = runCatching {
        outputFmt.format(inputFmt.parse(order.createdAt)!!)
    }.getOrDefault(order.createdAt.take(10))

    Surface(
        onClick        = onClick,
        shape          = MaterialTheme.shapes.large,
        color          = Surface,
        tonalElevation = 0.dp,
        modifier       = Modifier.fillMaxWidth(),
    ) {
        Column(modifier = Modifier.padding(16.dp)) {

            // Header
            Row(
                modifier              = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment     = Alignment.Top,
            ) {
                Column {
                    Text(
                        text       = "Pedido #${order.id}",
                        style      = MaterialTheme.typography.titleMedium,
                        fontWeight = FontWeight.Bold,
                        color      = TextPrimary,
                    )
                    Text(
                        text  = dateStr,
                        style = MaterialTheme.typography.bodySmall,
                        color = TextSecondary,
                    )
                }
                StatusBadge(status = order.status)
            }

            Spacer(Modifier.height(12.dp))

            // Preview ítems como chips
            Row(
                horizontalArrangement = Arrangement.spacedBy(6.dp),
                modifier              = Modifier.fillMaxWidth(),
            ) {
                val preview = order.items.take(3)
                preview.forEach { item ->
                    Surface(
                        color  = Surface2,
                        shape  = MaterialTheme.shapes.extraSmall,
                    ) {
                        Text(
                            text     = "${item.quantity}× ${item.productName}",
                            style    = MaterialTheme.typography.bodySmall.copy(fontSize = 11.sp),
                            color    = TextSecondary,
                            modifier = Modifier.padding(horizontal = 8.dp, vertical = 3.dp),
                            maxLines = 1,
                        )
                    }
                }
                if (order.items.size > 3) {
                    Text(
                        text  = "+${order.items.size - 3} más",
                        style = MaterialTheme.typography.bodySmall,
                        color = TextFaint,
                        modifier = Modifier.align(Alignment.CenterVertically),
                    )
                }
            }

            Spacer(Modifier.height(12.dp))
            HorizontalDivider(color = BorderLight, thickness = 0.5.dp)
            Spacer(Modifier.height(10.dp))

            // Footer
            Row(
                modifier              = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment     = Alignment.CenterVertically,
            ) {
                Text(
                    text  = "${order.numItems} producto${if (order.numItems != 1) "s" else ""}",
                    style = MaterialTheme.typography.bodySmall,
                    color = TextSecondary,
                )
                Text(
                    text       = "$${"%.2f".format(order.total)}",
                    style      = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.ExtraBold,
                    color      = Accent,
                )
            }
        }
    }
}